drivers/counter/counter_sam_tc.c

```diff
@@ -0,0 +1,404 @@
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	*ticks = tc_ch->TC_CV;
+
+	return 0;
+}
+
+static int counter_sam_tc_set_alarm(const struct device *dev,
+				    uint8_t chan_id,
+				    const struct counter_alarm_cfg *alarm_cfg)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+	uint32_t alarm_value;
+
+	__ASSERT_NO_MSG(alarm_cfg->callback != NULL);
+
+	if (chan_id >= MAX_ALARMS_PER_TC_CHANNEL) {
+		return -EINVAL;
+	}
+
+	uint32_t top_value = tc_ch->TC_RC != 0 ? tc_ch->TC_RC : UINT16_MAX + 1;
+
+	if (alarm_cfg->ticks > top_value) {
+		return -EINVAL;
+	}
+
+	if (data->alarm[chan_id].callback != NULL) {
+		return -EBUSY;
+	}
+
+	if (chan_id == 0) {
+		tc_ch->TC_IDR = TC_IDR_CPAS;
+	} else {
+		tc_ch->TC_IDR = TC_IDR_CPBS;
+	}
+
+	data->alarm[chan_id].callback = alarm_cfg->callback;
+	data->alarm[chan_id].user_data = alarm_cfg->user_data;
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) != 0) {
+		alarm_value = alarm_cfg->ticks;
+	} else {
+		alarm_value = (tc_ch->TC_CV + alarm_cfg->ticks) % top_value;
+	}
+
+	if (chan_id == 0) {
+		tc_ch->TC_RA = alarm_value;
+		/* Clear interrupt status register */
+		(void)tc_ch->TC_SR;
+		tc_ch->TC_IER = TC_IER_CPAS;
+	} else {
+		tc_ch->TC_RB = alarm_value;
+		/* Clear interrupt status register */
+		(void)tc_ch->TC_SR;
+		tc_ch->TC_IER = TC_IER_CPBS;
+	}
+
+	LOG_DBG("set alarm: channel %u, count %u", chan_id, alarm_value);
+
+	return 0;
+}
+
+static int counter_sam_tc_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+
+	if (chan_id >= MAX_ALARMS_PER_TC_CHANNEL) {
+		return -EINVAL;
+	}
+
+	if (chan_id == 0) {
+		tc_ch->TC_IDR = TC_IDR_CPAS;
+		tc_ch->TC_RA = 0;
+	} else {
+		tc_ch->TC_IDR = TC_IDR_CPBS;
+		tc_ch->TC_RB = 0;
+	}
+
+	data->alarm[chan_id].callback = NULL;
+	data->alarm[chan_id].user_data = NULL;
+
+	LOG_DBG("cancel alarm: channel %u", chan_id);
+
+	return 0;
+}
+
+static int counter_sam_tc_set_top_value(const struct device *dev,
+					const struct counter_top_cfg *top_cfg)
+{
+	struct counter_sam_dev_data *data = DEV_DATA(dev);
+	const struct counter_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Tc *tc = dev_cfg->regs;
+	TcChannel *tc_ch = &tc->TcChannel[dev_cfg->tc_chan_num];
+	int ret = 0;
+
+	for (int i = 0; i < MAX_ALARMS_PER_TC_CHANNEL; i++) {
+		if (data->alarm[i].callback) {
+			return -EBUSY;
+		}
+	}
+
+	/* Disable the compare interrupt */
+	tc_ch->TC_IDR = TC_IDR_CPCS;
+
+	data->top_cb = top_cfg->callback;
+	data->top_user_data = top_cfg->user_data;
+
+	tc_ch->TC_RC = top_cfg->ticks;
+
+	if ((top_cfg->flags & COUNTER_TOP_CFG_DONT_RESET) &&
```

Comment:
i have no idea how timer works but seems to me that there is the same behavior when  `DONT_RESET` option is set and current tick is smaller than new value and case when `DONT_RESET` option is off. Both lands in `c_ch->TC_CCR = TC_CCR_SWTRG`.